File: service/bridge.go
package service

import (
	"encoding/json"
	"fmt"
	"log"

	"github.com/etcd-io/etcd/clientv3"
	"github.com/vicanso/etcd-dashboard/etcd"
	"github.com/zserge/lorca"
)

// BridgeBind bridge bind
func BridgeBind(ui lorca.UI) {
	c := &Config{}
	ui.Bind("BConfigSet", c.Set)
	ui.Bind("BConfigGet", c.Get)
	ui.Bind("BInitEtcdClient", initEtcdClient)
	ui.Bind("BGetKeysWithPrefix", getKeysWithPrefix)
	ui.Bind("BGetValueByKey", getValueByKey)
	ui.Bind("BPutValueByKey", putValueByKey)
	ui.Bind("BDelValueByKey", delValueByKey)
}

func errorNotify(err error) {
	if err == nil {
		return
	}
	log.Printf("error:%v", err)
}

func initEtcdClient(name, config string) {
	c := clientv3.Config{}
	err := json.Unmarshal([]byte(config), &c)
	if err != nil {
		errorNotify(err)
		return
	}
	_, err = etcd.InitClient(name, c)
	fmt.Println(c)
	if err != nil {
		errorNotify(err)
	}
	return
}

func getKeysWithPrefix(name, prefix string) (keys []string, err error) {
	keys, err = etcd.GetKeys(name, prefix, &etcd.OpOptions{
		WithPrefix: true,
	})
	return
}

func getValueByKey(name, key string) (value string, err error) {
	resp, err := etcd.Get(name, key, nil)
	if err != nil {
		return
	}
	if len(resp.Kvs) != 0 {
		value = string(resp.Kvs[0].Value)
	}
	return
}

// putValueByKey put value by key
func putValueByKey(name, key, value string) (result string, err error) {
	_, err = etcd.Put(name, key, value, nil)
	result = "fail"
	if err != nil {
		return
	}
	result = "ok"
	return
}

// delValueByKey delete value by key
func delValueByKey(name, key string) (result string, err error) {
	_, err = etcd.Del(name, key, nil)
	result = "fail"
	if err != nil {
		return
	}
	result = "ok"
	return
}

File: dashboard/src/bridge.js
const ready = !!window["BConfigSet"];

const debug = function(...args) {
  console.info(...args)
}

const bridge = {
  SetWidowSize: async function(width, height) {
    if (!ready) {
      return;
    }
    window.BConfigSet("width", width);
    window.BConfigSet("height", height);
  },
  addConnection: async function(connection) {
    if (!ready) {
      return;
    }
    const found = await this.getConnection();
    if (!found) {
      throw new Error('The name is exists, please use update');
    }
    // const connections = await this.getConnections();
    // let found = null;
    // connections.forEach((item) => {
    //   if (item.name === connection.name) {
    //     found = item;
    //   }
    // });
    // if (found) {
    //   found = Object.assign(found, connection);
    // } else {
      // connections.push(connection);
    // }
    const connections = await this.getConnections(); 
    connections.push(connection);
    await window.BConfigSet("connections", JSON.stringify(connections));
  },
  getConnections: async function() {
    if (!ready) {
      return [];
    }
    const values = await window.BConfigGet("connections");
    if (!values) {
      return [];
    }
    const data = JSON.parse(values);
    debug("connections:", data);
    return data
  },
  getConnection: async function(name) {
    const connections = await this.getConnections(); 
    let found = null;
    connections.forEach((item) => {
      if (item.name === name) {
        found = item;
      }
    });
    return found;
  },
  initConnection: async function(name) {
    const found = await this.getConnection(name) ;
    if (!found) {
      throw new Error('can not find the config for the connection')
    }
    const config = {
      endpoints: found.endPoints.split(','),
    };
    if (found.username) {
      config.username = found.username;
      config.password = found.password;
    }
    await window.BInitEtcdClient(found.name, JSON.stringify(config));
  },
  getKeysWithPrefix: async function(name, prefix) {
    const keys = await window.BGetKeysWithPrefix(name, prefix);
    debug("keys:", keys);
    return keys
  },
  getValueByKey: async function(name, key) {
    const value = await window.BGetValueByKey(name, key);
    debug("get key: %s, value:%s", key, value);
    return value;
  },
  putValueByKey: async function(name, key, value) {
    await window.BPutValueByKey(name, key, value);
    debug("set key: %s, value:%s success", key, value);
  },
  delValueByKey: async function(name, key) {
    await window.BDelValueByKey(name, key);
    debug("del key: %s success", key);
  }
}

export default bridge;